assignment3/Q1.cpp

```diff
@@ -0,0 +1,125 @@
+#include <iostream>
+#include <vector>
+#include <list>
+using namespace std;
+
+
+
+list<string> searchFromPosition(int rows, int cols, char**board, bool **used, vector<string> dict, int x, int y, string word) {
+
+    list<string> part;
+
+    //unavailable position
+    if(board[x][y] == '0' || (used[x-1][y-1] == true))
+        return part;
+
+    word = word + board[x][y];
+    //present string is not prefix
+    if(!dict.isPrefix(word)) {
+        return part;
+    }
+
+    //mark as used
+    used[x-1][y-1] = true;
+    if(dict.isWord(word))
+        part.insert(part.begin(), word);
+
+
+    //check neighbours
+    list<string> l;
+    l = searchFromPosition(rows, cols, board, used, dict, x-1, y-1, word);
+    part.merge(l);
+    l = searchFromPosition(rows, cols, board, used, dict, x-1, y, word);
+    part.merge(l);
+    l = searchFromPosition(rows, cols, board, used, dict, x-1, y+1, word);
+    part.merge(l);
+    l = searchFromPosition(rows, cols, board, used, dict, x, y-1, word);
+    part.merge(l);
+    l = searchFromPosition(rows, cols, board, used, dict, x, y+1, word);
+    part.merge(l);
+    l = searchFromPosition(rows, cols, board, used, dict, x+1, y-1, word);
+    part.merge(l);
+    l = searchFromPosition(rows, cols, board, used, dict, x+1, y, word);
+    part.merge(l);
+    l = searchFromPosition(rows, cols, board, used, dict, x+1, y+1, word);
+    part.merge(l);
+
+
+    //delete duplicates
+    part.unique();
+    return part;
+}
+
+
+list<string> wordsSearch(int rows, int cols, char **board, vector<string> dict) {
+
+
+    list<string> result;
+
+    //create array with borders (for simpler checking bounds)
+    char **board_w_borders = new char *[rows+2];
+    for(int i = 0; i < rows+2; i++) {
+        board_w_borders[i] = new char[cols+2];
+    }
+    //fill inside
+    for(int i = 1; i < rows+1; i++) {
+        for(int j = 1; j < cols+1; j++) {
+            board_w_borders[i][j] = board[i-1][j-1];
+        }
+    }
+    //fill borders - I assume that words consist only of letters, not numbers
+    for(int i = 0; i < rows + 2; i++) {
+        board_w_borders[i][0] = '0';
```

Comment:
It is ok to assume that words consist only letters.
But I think a better value for this will be `'\0'`, which is the 'null' character, which essentially means "nothing".
This is the real "zero" you are looking for, and not the zero character ;)
This way you don't need to assume anything. Also it seems less random than choosing the character `'0' ` or any other non-null character.

You can save it to a constant, to avoid having 'magic' values in your code:
`const bool outOfBorder = '\0';`;
and then the following would be very clear to understand:
`if (board[x][y] == outOfBorder) return;`